metadata/pkg/drivers/aws/aws.go

```diff
@@ -43,10 +50,122 @@ type AwsAdapter struct {
+	}
+	if meta.ServerSideEncryption != nil {
+		obj.ServerSideEncryption = *meta.ServerSideEncryption
+	}
+	if meta.VersionId != nil {
+		obj.VersionId = *meta.VersionId
+	}
+	obj.ObjectType = *meta.ContentType
+	if meta.Expires != nil {
+		expiresTime, err := time.Parse(time.RFC3339, *meta.Expires)
+		if err != nil {
+			log.Errorf("unable to parse given string to time type. error: %v. skipping ExpiresDate field\n", err)
+		} else {
+			obj.ExpiresDate = expiresTime
+		}
+	}
+	if meta.ReplicationStatus != nil {
+		obj.ReplicationStatus = *meta.ReplicationStatus
+	}
+	objectArray[i] = obj
+}
+
+func ObjectList(sess *session.Session, bucket *model.MetaBucket) error {
+	svc := s3.New(sess, aws.NewConfig().WithRegion(bucket.Region))
+	output, err := svc.ListObjectsV2(&s3.ListObjectsV2Input{Bucket: &bucket.Name})
+	if err != nil {
+		log.Errorf("unable to list objects in bucket %v. failed with error: %v\n", bucket.Name, err)
+		return err
+	}
+
+	numObjects := len(output.Contents)
+
+	objectArray := make([]model.MetaObject, numObjects)
+	wg := sync.WaitGroup{}
+	for i, object := range output.Contents {
+		wg.Add(1)
+		go iterateObjects(i, object, svc, bucket.Name, objectArray, &wg)
+	}
+	wg.Wait()
+	bucket.NumberOfObjects = numObjects
+	bucket.Objects = objectArray
+	return nil
+}
+
+func iterateBuckets(i int, bucket *s3.Bucket, sess *session.Session, bucketArray []model.MetaBucket, wg *sync.WaitGroup) {
+	defer wg.Done()
+	buck := model.MetaBucket{}
+	buck.CreationDate = *bucket.CreationDate
+	buck.Name = *bucket.Name
+	svc := s3.New(sess)
+	loc, err := svc.GetBucketLocation(&s3.GetBucketLocationInput{Bucket: bucket.Name})
+	if err != nil {
+		log.Errorf("unable to get bucket location. failed with error: %v\n", err)
+	} else {
+		buck.Region = *loc.LocationConstraint
+	}
+	newSvc := s3.New(sess, aws.NewConfig().WithRegion(buck.Region))
+	tags, err := newSvc.GetBucketTagging(&s3.GetBucketTaggingInput{Bucket: bucket.Name})
+
+	if err != nil && !strings.Contains(err.Error(), "NoSuchTagSet") {
+		log.Errorf("unable to get bucket tags. failed with error: %v\n", err)
+	} else {
+		tagset := make(map[string]string)
+		for _, tag := range tags.TagSet {
+			tagset[*tag.Key] = *tag.Value
+		}
+		buck.BucketTags = tagset
+	}
+	err = ObjectList(sess, &buck)
+	if err != nil {
+		log.Errorf("error while collecting object metadata for bucket %v. failed with error: %v\n", buck.Name, err)
+	}
+	bucketArray[i] = buck
+}
+
+func BucketList(sess *session.Session) ([]model.MetaBucket, error) {
+	svc := s3.New(sess)
+
+	output, err := svc.ListBuckets(&s3.ListBucketsInput{})
+	if err != nil {
+		log.Errorf("unable to list buckets. failed with error: %v\n", err)
+		return nil, err
+	}
+	numBuckets := len(output.Buckets)
+	bucketArray := make([]model.MetaBucket, numBuckets)
+	wg := sync.WaitGroup{}
+	for i, bucket := range output.Buckets {
+		wg.Add(1)
+		go iterateBuckets(i, bucket, sess, bucketArray, &wg)
```

Comment:
please don't use go routine here.(mostly it's nested go routines usage.)

Reply:
done